### ISSUE TITLE
Workflow of modifying CPython in Pyodide

### ISSUE THREAD

**Issue by rth** [pyodide/pyodide#2710]
Related to a recent discussion with @antocuni we should probably document better the recommended workflow for working on a modifed CPython as part of Pyodide. To some extent, this workflow also applies to other patched packages.

By default, the build system would indeed [download the tar.gz](https://github.com/pyodide/pyodide/blob/main/cpython/Makefile#L11), extract to `cpython/build/` and apply patches.

Potentially one could just modify files in this build directory and re-run make. It should not overwrite changes unless one runs `make -C cpython clean`. However, the problem is then it's difficult to create patches or generally track changes.

Another solution that we have been using more lately, is to 
1. clone the CPython repo via your Github fork
2.  checkout the version we use now (3.10.2)
3.  create a new branch and apply all patches in `cpython/patches` on this branch.
4.  Then work in this branch, which has the advantage that new commits can be easily exported as patches as documented [here](https://pyodide.org/en/latest/development/contributing.html#migrating-patches)

To make things faster for development and avoid exporting patches each time, one can likely symlink the Cpython working tree to `pyodide/cpython/build` though, one needs to run,
 ```
touch pyodide/cpython/build/.patched
```
so it [wouldn't be overwritten](https://github.com/pyodide/pyodide/blob/main/cpython/Makefile#L86). I think it should work, but personally, I have not tried this last part. 

Maybe @hoodmane who has worked more on updating CPython more recently has also some thoughts on this workflow.

**Comment by rth** [comment]
In terms of documentation, we have https://pyodide.org/en/latest/development/contributing.html#migrating-patches but maybe we should re-phrase it to be not only about migrating patches. And also indicate how one can avoid serializing any new changes as patches during development (as mentioned above) and probably mention interactive git rebase which is really helpful for managing git history and creating commits that can later be exported as patches.

**Comment by antocuni** [comment]
Thank you for pointing me in the right direction. 
I noticed that after modifying CPython, a simple `make` doesn't recompile it. I managed to find a way to have a working build/modify cpython/build cycle but it looks a bit suboptimal and probably slower than necessary by doing the following:
```
$ make # first make
...

SUCCESS!

$ emacs cpython/build/Python-3.10.2/Python/sysmodule.c  # modify CPython
$ touch cpython/build/Python-3.10.2/Makefile
$ make -C cpython
$ make
```

The `touch Makefile` is needed else the following `make -C cpython` doesn't recompile anything; and if I do a simple `make` instead of `make -C cpython && make`, pyodide still uses the old build of CPython without getting my modifications.

In general, it would be very useful to have some documentation about the build process, because it seems to be complex and not so easy to follow.

---

> Another solution that we have been using more lately, is to
> 
> 1. clone the CPython repo via your Github fork
> 2. checkout the version we use now (3.10.2)
> 3. create a new branch and apply all patches in `cpython/patches` on this branch.
> 4. Then work in this branch, which has the advantage that new commits can be easily exported as patches as documented [here](https://pyodide.org/en/latest/development/contributing.html#migrating-patches)

have you considered to have an official pyodide fork of CPython (living e.g. at github.com/pyodide/cpython or so) and to use git submodules to integrate it with the pyodide repo?
This would make things much easier, and "migrating patches" would be as easy as rebasing the branch on top of new CPython versions.

**Comment by antocuni** [comment]
> > The touch Makefile is needed else the following make -C cpython doesn't recompile anything
> 
> Hmm, yes, good point. I can't say I am very knowledgeable about makefiles, so it has been a bit of trial and error so far for such things.

same here, I'm always confused by Makefiles :)

> > In general, it would be very useful to have some documentation about the build process, because it seems to be complex and not so easy to follow.
> 
> Agreed. We have been trying to move away from Makefiles as the top level API in favor of building scripts in pyodide-build, but it's been a slow process. There are some docs for [package building](https://pyodide.org/en/latest/development/new-packages.html) but indeed less so for CPython. Though personally I think that if we can move CPython to be a package (although with some specifics) as is it the case in conda, it would simplify things in terms of build.

I was more thinking of a higher level overview of what's happening: I suppose that somehow the process is roughly:

1. download/unpack the .tar.gz
2. apply patches
3. make CPython using the emscripten SDK, the result is <something>.js which is in <dir>
4. magic happens
5. the final result is <something> which is in <dir>
 
The points 3-5 are a bit obscure to me :)

> > have you considered to have an official pyodide fork of CPython
> 
> We have considered it for packages (e.g. numpy, scipy) that also used to have lots of patches (though less for numpy recently). The question is how one would contribute to it. Right now one makes a single PR to this repo with patches, which is a bit cumbersome but overall works. If there is a fork that we include as a submodule, the PR would need to be done there. The issue is that we then need to run the full pyodide build / tests pipeline to make sure the changes work (maybe via a second PR to this repo?)

Ah, I understand the issue. I think that the plan is to apply all the patches upstream to be able to use a vanilla CPython anyway, at some point?

> Also, the situation is that some of these patches are ideally contributed upstream, and would disappear in time. And so then either it means that in the fork we add/remove these patches as commits, which will be more difficult to backport to future Python versions (since each commit needs then to be ported as far as I understand). Or we rewrite git history in the fork to remove the added patch, but then we have no longer a source of truth for git history.

I think this would automatically be handled when you do a rebase. But yes, this would mean to rewrite git history.

> Also so far I haven't had a very good experience with git submodules (e.g. the diff on Github is not very useful when updating it), but maybe this could indeed be a good use case for them.

I admit I don't have much experience either

**Comment by hoodmane** [comment]
> The touch Makefile is needed else the following make -C cpython doesn't recompile anything; and if I do a simple make instead of make -C cpython && make, pyodide still uses the old build of CPython without getting my modifications.

There are several bugs in the Makefiles. The root Makefile runs rollup twice for no reason and seems to almost always think that the build is stale. The CPython Makefile as you note requires you to touch the deeper Makefile to get it to rebuild, and there is also a problem that it injects a libffi header into the Python build directory which gets deleted in some partial rebuilds causing linker errors. I think these are the three most annoying problems.

**Comment by hoodmane** [comment]
> I think that the plan is to apply all the patches upstream to be able to use a vanilla CPython anyway, at some point?

Yeah all 12 of the current patches have been upstreamed into Python 3.11. I think we may want to add a new patch for #2142 but I think that will be easy to upstream when we get it working.
